TSPL.docc/LanguageGuide/TheBasics.md

```diff
@@ -86,6 +86,72 @@ because the maximum value never changes.
 The current login attempt counter is declared as a variable,
 because this value must be incremented after each failed login attempt.
 
+If a stored value in your code won't change,
+always declare it as a constant with the `let` keyword.
+Use variables only for storing values that need to be able to change.
+
+When you declare a constant or a variable,
+you can give it a value as part of that declaration,
+like the examples above.
+Alternately,
+you can assign its initial value later in the program,
+as long as it's guaranteed to have a value
+before the first time you read from it.
+
+```swift
+var environment = "development"
+let maximumNumberOfLoginAttempts: Int
+// maximumNumberOfLoginAttempts has no value yet.
+
+if environment == "development" {
+    maximumNumberOfLoginAttempts = 100
+} else {
+    maximumNumberOfLoginAttempts = 10
+}
+// Now maximumNumberOfLoginAttempts has a value, and can be read.
+```
+
+<!--
+  - test: `constantsWithDeferredInitialization`
+
+  ```swifttest
+  -> var environment = "development"
+  -> let maximumNumberOfLoginAttempts: Int
+  -> if environment == "development" {
+         maximumNumberOfLoginAttempts = 100
+     } else {
+         maximumNumberOfLoginAttempts = 10
+     }
+  >> print(maxNumberOfLoginAttempts)
+  << 100
+  ```
+-->
+
+In this example,
+the maximum number of login attempts is constant,
+and its value depends on the environment.
+In the development environment,
+it has a value of 100;
+in any other environment, its value is 10.
+Both branches of the `if` statement
+initialize `maximumNumberOfLoginAttempts` with some value,
+which guarantees that the constant always gets a value.
+
+When you define a constant or variable
+without giving it a value,
+Swift analyzes your code at compile time
+and confirms that all possible paths of execution
+are guaranteed to set a value before reading a value.
+This analysis is called *definitive initialization*.
+
+> Note:
+> Definitive initialization
+> can't construct proofs that require domain knowledge,
+> and its ability to track state across conditionals is finite.
+> If you can determine that a value is always set,
+> but Swift can't prove this is the case,
+> try simplifying the conditionals or use a variable instead.
```

Comment:
@chuckdude I rewrote this note once, but I'm still not quite happy with it.  Essentially, the full generality of DI reduces to the halting problem, so it's provably impossible for the compiler to always detect.  Maybe this detail should move to the reference.

Reply:
Moved DI details to the reference.  Now I'm debating whether an xref is even needed.